pkg/kv/postgres/store.go

```diff
@@ -309,54 +317,38 @@ func (e *EntriesIterator) Next() bool {
 	if e.err != nil {
 		return false
 	}
-	e.prev = e.entry
-	e.entry = nil
-	if !e.rows.Next() {
-		if e.prev == nil {
+	e.currEntryIdx++
+
+	if e.currEntryIdx == len(e.entries) {
+		if e.currEntryIdx == 0 {
 			return false
 		}
-		// A previous row was read successfully, and no next row is found - this could
-		// be the end of a page, so trying to get the next page (another call to scanInternal)
-		// If there will still be no next row, it is the end of the scan
-		tmpIter, err := e.store.scanInternal(e.ctx, e.prev.PartitionKey, e.prev.Key, false)
+		tmpIter, err := e.store.scanInternal(e.ctx, e.entries[e.currEntryIdx-1].PartitionKey, e.entries[e.currEntryIdx-1].Key, false)
 		if err != nil {
 			e.err = fmt.Errorf("scan paging: %w", err)
 			return false
 		}
-		e.rows = tmpIter.rows
-
-		// if there is still no next row, the end of the scan is reached
-		if !e.rows.Next() {
+		if len(tmpIter.entries) == 0 {
 			return false
 		}
+		e.entries = tmpIter.entries
+		e.currEntryIdx = 0
 	}
-	var ent kv.Entry
-	if err := e.rows.Scan(&ent.PartitionKey, &ent.Key, &ent.Value); err != nil {
-		e.err = fmt.Errorf("%s: %w", err, kv.ErrOperationFailed)
-		return false
-	}
-	e.entry = &ent
+
 	return true
 }
 
 func (e *EntriesIterator) Entry() *kv.Entry {
-	return e.entry
+	return &e.entries[e.currEntryIdx]
```

Comment:
Can this panic? Isn't `nil` a better option?

Reply:
Good point - it certainly can.
Fixed